src/server/snapshot.cc

```diff
@@ -108,32 +108,39 @@ void SliceSnapshot::StartIncremental(Context* cntx, LSN start_lsn) {
 
 // Called only for replication use-case.
 void SliceSnapshot::Finalize() {
+  DVLOG(1) << "Finalize Snapshot";
   DCHECK(db_slice_->shard_owner()->IsMyThread());
   DCHECK(journal_cb_id_);
 
   // Wait for serialization to finish in any case.
   snapshot_fb_.JoinIfNeeded();
 
+  if (unregister_journal_) {
+    return;
+  }
   auto* journal = db_slice_->shard_owner()->journal();
   serializer_->SendJournalOffset(journal->GetLsn());
+  unregister_journal_ = true;
   journal->UnregisterOnChange(journal_cb_id_);
-  journal_cb_id_ = 0;
+
   PushSerializedToChannel(true);
   CloseRecordChannel();
 }
 
-void SliceSnapshot::StopChannel() {
-  VLOG(1) << "SliceSnapshot::StopChannel";
+// Called only for replication use-case.
```

Comment:
Seems that Finalize and Cancel are almost the same. We can unify them into the same function `FinalizeReplication(bool cancel)`.